File: backend/utils/timezone.py
"""
Timezone utilities for converting UTC to IST (Indian Standard Time)
"""
from datetime import datetime, timezone, timedelta
from typing import Optional

# IST timezone offset (UTC+5:30)
IST = timezone(timedelta(hours=5, minutes=30))

def now_ist() -> datetime:
    """Get current time in IST timezone"""
    return datetime.now(IST)

def utc_to_ist(utc_dt: datetime) -> datetime:
    """Convert UTC datetime to IST timezone"""
    if utc_dt.tzinfo is None:
        # Assume naive datetime is UTC
        utc_dt = utc_dt.replace(tzinfo=timezone.utc)
    return utc_dt.astimezone(IST)

def ist_to_utc(ist_dt: datetime) -> datetime:
    """Convert IST datetime to UTC timezone"""
    if ist_dt.tzinfo is None:
        # Assume naive datetime is IST
        ist_dt = ist_dt.replace(tzinfo=IST)
    return ist_dt.astimezone(timezone.utc)

def format_ist_timestamp(dt: Optional[datetime] = None) -> str:
    """Format datetime as IST timestamp string (HH:MM:SS format)"""
    if dt is None:
        dt = now_ist()
    elif dt.tzinfo is None:
        # Assume naive datetime is UTC and convert to IST
        dt = dt.replace(tzinfo=timezone.utc).astimezone(IST)
    elif dt.tzinfo != IST:
        # Convert to IST if different timezone
        dt = dt.astimezone(IST)
    
    return dt.strftime("%H:%M:%S")

def get_ist_datetime_for_db() -> datetime:
    """Get current IST datetime for database storage (as naive datetime)"""
    return now_ist().replace(tzinfo=None)

File: frontend/src/utils/timezone.ts
/**
 * Timezone utilities for handling IST (Indian Standard Time)
 */

// IST offset: UTC+5:30
const IST_OFFSET_HOURS = 5;
const IST_OFFSET_MINUTES = 30;

/**
 * Convert a date to IST timezone
 */
export const toIST = (date: Date | string): Date => {
  const d = typeof date === 'string' ? new Date(date) : date;
  
  // If the date doesn't have timezone info, assume it's already in IST from the backend
  if (typeof date === 'string' && !date.includes('Z') && !date.includes('+') && !date.includes('-')) {
    return d;
  }
  
  // Convert UTC to IST
  const utc = d.getTime() + (d.getTimezoneOffset() * 60000);
  const istTime = new Date(utc + (IST_OFFSET_HOURS * 3600000) + (IST_OFFSET_MINUTES * 60000));
  
  return istTime;
};

/**
 * Format a date/timestamp for display with IST consideration
 */
export const formatISTTimestamp = (timestamp: string | Date, formatStr: string = 'MMM dd, yyyy HH:mm:ss'): string => {
  // Import format function from date-fns here to avoid circular dependencies
  const { format } = require('date-fns');
  
  const date = toIST(timestamp);
  return format(date, formatStr) + ' IST';
};

/**
 * Get current IST time
 */
export const nowIST = (): Date => {
  const now = new Date();
  return toIST(now);
};